### PR TITLE
Knative error extensions

### LINKED ISSUE #2340: Failed events are not enhanced with extension attributes
**Describe the bug**
Failed events are not enhanced with extension attributes prior to forwarding to the deadLetterSink as described here: 
https://knative.dev/docs/eventing/event-delivery/#configuring-channel-event-delivery
(knativeerrordest, knativeerrorcode, knativeerrordata) 

 **Expected behavior**
I expect failed events to be enhanced with extension attributes.

 **Knative release version**
1.5

 **Additional context**
Sending event to deadLetterSink is kind of pointless without some context (where this event was meant to be delivered; what happened). There is not really much i can do with such event - i don't know which subscriber failed to consume it and why

### PR DESCRIPTION
**Pull request by egorlepa**
Fixes #2340

### CODE REVIEW

**Comment by matzew** [comment]
@egorlepa thanks for putting this PR together, I looks nice!

I will review this tomorrow. Glad to see your contribution here :tada:

**Comment by egorlepa** [comment]
@matzew ok, cool. i'll fix e2es tomorrow

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java [hunk 1]
```diff
@@ -257,9 +266,61 @@ private void onSubscriberFailure(final Throwable failure,
     incrementEventCount(response, recordContext);
     recordDispatchLatency(response, recordContext);
 
-    dlsSender.apply(recordContext.getRecord())
-      .onSuccess(v -> onDeadLetterSinkSuccess(recordContext, finalProm))
-      .onFailure(ex -> onDeadLetterSinkFailure(recordContext, ex, finalProm));
+    // enhance event with extension attributes prior to forwarding to the dead letter sink
+    ConsumerRecordContext transformedRecordContext = errorTransform(recordContext, response);
```
Comment (matzew) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:270]: I'd prefer using `final` as much as possible, to express objects are not mutable.

See other classes for examples. That would be really my only "nit" comment on the code :smile:

Comment (egorlepa) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:270]: ok, no problem. i mainly use golang and kotlin nowadays, almost forgot how to talk in java :) kotlin has nice val/var for this purpose

**Comment by matzew** [comment]
@egorlepa it would be nice if we had a little e2e test for this in here: https://github.com/knative-sandbox/eventing-kafka-broker/tree/main/test/e2e_new

File: test/e2e_new/dls_extenisons_test.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
+//go:build e2e
+// +build e2e
```
Comment (matzew) [inline test/e2e_new/dls_extenisons_test.go:2]: `dls_extenisons_test.go`: there is a little typo in `extensions`

File: test/e2e_new/dls_extensions_test.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+	))
+	f.Setup("install trigger", trigger.Install(
+		triggerName,
+		brokerName,
+		trigger.WithSubscriber(svc.AsKReference(sinkName), ""),
+		trigger.WithDeadLetterSink(svc.AsKReference(deadLetterSinkName), ""),
+	))
+	f.Setup("trigger is ready", trigger.IsReady(triggerName))
+
+	f.Setup("install source", eventshub.Install(
```
Comment (pierDipi) [inline test/e2e_new/dls_extensions_test.go:206]:
```suggestion
	f.Requirement("install source", eventshub.Install(
```
Same reasoning.

File: test/reconciler-tests.sh [hunk 4]
```diff
@@ -27,7 +27,7 @@ if [ "${EVENTING_KAFKA_BROKER_CHANNEL_AUTH_SCENARIO:-""}" != "" ]; then
   success
 fi
 
-go_test_e2e -timeout=1h ./test/e2e_new/... || fail_test "E2E (new) suite failed"
+go_test_e2e -timeout=1h -run TestDeadLetterSinkExtensions ./test/e2e_new/... || fail_test "E2E (new) suite failed"
```
Comment (matzew) [inline test/reconciler-tests.sh:30]: can we revert this change?

**Comment by matzew** [comment]
I am getting

```
panic: non-positive interval for NewTicker

goroutine 10 [running]:
time.NewTicker(0x1b2c400)
	time/tick.go:24 +0x10f
knative.dev/reconciler-test/pkg/eventshub/sender.Start({0x1b2c400, 0xc000764780}, 0x61f759)
	knative.dev/reconciler-test@v0.0.0-20220706134107-6ffd1789775c/pkg/eventshub/sender/sender.go:172 +0x56a
main.main.func4({0x1b2c400, 0xc000764780}, 0xc000490f98)
	knative.dev/reconciler-test@v0.0.0-20220706134107-6ffd1789775c/cmd/eventshub/main.go:46 +0x25
knative.dev/reconciler-test/pkg/eventshub.startEventGenerators.func1()
```

locally...

File: test/reconciler-tests.sh (hunk 4 above)
Comment (pierDipi) [reply test/reconciler-tests.sh]: I suggested this to get feedback faster from CI, we will revert it once it passes

**Comment by egorlepa** [comment]
yep, SendMultipleEvents duration can't be <1. how am i supposed to know that :)

File: test/e2e_new/dls_extensions_test.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+		trigger.WithDeadLetterSink(svc.AsKReference(deadLetterSinkName), ""),
+	))
+	f.Setup("trigger is ready", trigger.IsReady(triggerName))
+
+	f.Requirement("install source", eventshub.Install(
+		sourceName,
+		eventshub.StartSenderToResource(broker.GVR(), brokerName),
+		eventshub.InputEventWithEncoding(ev, cloudevents.EncodingBinary),
+		eventshub.AddSequence,
+		eventshub.SendMultipleEvents(1, 100*time.Millisecond),
```
Comment (matzew) [inline test/e2e_new/dls_extensions_test.go:95]: `time` import is missing

Comment (egorlepa) [reply test/e2e_new/dls_extensions_test.go:96]: ah yeah.

File: test/e2e_new/dls_extensions_test.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	))
+
+	f.Assert("knativeerrordest, knativeerrorcode, knativeerrordata added", assertEnhancedWithKnativeErrorExtensions(
+		deadLetterSinkName,
```
Comment (matzew) [inline test/e2e_new/dls_extensions_test.go:216]:
```
=== RUN   TestDeadLetterSinkExtensions/SubscriberReturnedErrorSmallData/Assert/[Stable/MUST]knativeerrordest,_knativeerrorcode,_knativeerrordata_added
=== CONT  TestDeadLetterSinkExtensions/SubscriberReturnedErrorSmallData/Assert/[Stable/MUST]knativeerrordest,_knativeerrorcode,_knativeerrordata_added
    event_info_store.go:157: Timeout waiting for at least 1 matches.
        Error: FAIL MATCHING: saw 0/1 matching events.
        - Store-
        Pod 'dls-fxaxwvpa' in namespace 'test-xsnfgztx'
        - Recent events -
        0 events seen, last 0 events (total events seen 19, events ignored 19):
```

there are no events routed to the `deadLetterSink` in this test case

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java [hunk 7]
```diff
@@ -258,9 +267,61 @@ private void onSubscriberFailure(final Throwable failure,
...
+    final var transformedRecordContext = errorTransform(recordContext, response);
+
+    dlsSender.apply(transformedRecordContext.getRecord())
+      .onSuccess(v -> onDeadLetterSinkSuccess(transformedRecordContext, finalProm))
+      .onFailure(ex -> onDeadLetterSinkFailure(transformedRecordContext, ex, finalProm));
+  }
+
+  private ConsumerRecordContext errorTransform(final ConsumerRecordContext recordContext, @Nullable final HttpResponse<?> response) {
+    final var destination = resourceContext.getEgress().getDestination();
+    if (response == null) {
```
Comment (matzew) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: when a sink returns something like this:

```golang
	log.Print("Got request, return a bad one")
	w.WriteHeader(http.StatusUnprocessableEntity)
	w.Write([]byte("Bad error"))
```

The `response` here is actually `null`.

Comment (matzew) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: Seeing something:

```
{"@timestamp":"2022-08-17T15:55:01.886Z","@version":"1","message":"Failed to handle subscriber response topic=knative-broker-default-my-demo-kafka-broker partition=7 offset=0","logger_name":"dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl","thread_name":"vert.x-worker-thread-16","level":"ERROR","level_value":40000,"stack_trace":"java.lang.IllegalStateException: Unable to decode response: unknown encoding and non empty response\n\tat dev.knative.eventing.kafka.broker.dispatcher.impl.BaseResponseHandler.handle(BaseResponseHandler.java:67)\n\tat dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl.lambda$composeSenderAndSinkHandler$8(RecordDispatcherImpl.java:369)\n\tat io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38)\n\tat io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262)\n\tat io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)\n\tat io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)\n\tat io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)\n\tat io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)\n\tat io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88)\n\tat io.vertx.core.impl.future.Composition.onSuccess(Composition.java:43)\n\tat io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)\n\tat io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)\n\tat io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)\n\tat dev.knative.eventing.kafka.broker.dispatcher.impl.http.WebClientCloudEventSender.lambda$send$5(WebClientCloudEventSender.java:175)\n\tat io.vertx.core.impl.future.FutureImpl$1.onSuccess(FutureImpl.java:91)\n\tat io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262)\n\tat io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)\n\tat io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)\n\tat io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)\n\tat io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)\n\tat io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)\n\tat io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)\n\tat io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:400)\n\tat io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:387)\n\tat io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:365)\n\tat io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:332)\n\tat io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:294)\n\tat io.vertx.ext.web.client.impl.HttpContext.lambda$null$8(HttpContext.java:550)\n\tat io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)\n\tat io.vertx.core.impl.WorkerContext.lambda$run$1(WorkerContext.java:83)\n\tat io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\nCaused by: io.cloudevents.rw.CloudEventRWException: Could not parse. Unknown encoding. Invalid content type or spec version\n\tat io.cloudevents.rw.CloudEventRWException.newUnknownEncodingException(CloudEventRWException.java:201)\n\tat io.cloudevents.core.message.impl.MessageUtils.parseStructuredOrBinaryMessage(MessageUtils.java:80)\n\tat io.cloudevents.http.vertx.VertxMessageFactory.createReader(VertxMessageFactory.java:46)\n\tat io.cloudevents.http.vertx.VertxMessageFactory.createReader(VertxMessageFactory.java:89)\n\tat dev.knative.eventing.kafka.broker.dispatcher.impl.BaseResponseHandler.handle(BaseResponseHandler.java:51)\n\t... 34 common frames omitted\n","topic":"knative-broker-default-my-demo-kafka-broker","partition":7,"offset":0}

```

Comment (pierDipi) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: To get the response from a failure we can use this logic:
```java
if (failure instanceof ResponseFailureException) {
  response = ((ResponseFailureException) failure).getResponse();
}
```

Comment (matzew) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: Ok, I will go that route - thx

Comment (matzew) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: It's actually `IllegalStateException`. The above log statement has this extracted snippet:


```
java.lang.IllegalStateException: Unable to decode response: unknown encoding and non empty response\n\tat dev.knative.eventing.kafka.broker.dispatcher.impl.BaseResponseHandler.handl......
```

Comment (matzew) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: After we have the fix #2521 in,

(and I added the cast to `ResponseFailureException`), I am getting:

```
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: demo
  source: /my/curl/command
  id: 4711
  datacontenttype: application/json
Extensions,
  knativeerrorcode: 422
  knativeerrordata: Bad error
  knativeerrordest: http://error-sender.default.svc.cluster.local
  smartevent: super-duper-event-extension
Data,
  {
    "message": "Hallo World"
  }


```

on the DeadletterSink

/cc @pierDipi

Comment (matzew) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/RecordDispatcherImpl.java:280]: @egorlepa I will PR you a diff to your PR, than we get this landed !

File: test/e2e_new/dls_extensions_test.go [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+		eventshub.StartSenderToResource(broker.GVR(), brokerName),
+		eventshub.InputEventWithEncoding(ev, cloudevents.EncodingBinary),
+		eventshub.AddSequence,
+		eventshub.SendMultipleEvents(1, 100*time.Millisecond),
+	))
+
+	f.Assert("knativeerrordest & knativeerrorcode added", assertEnhancedWithKnativeErrorExtensions(
+		deadLetterSinkName,
+		func(ctx context.Context) cetest.EventMatcher {
+			dlsAddress, _ := svc.Address(ctx, deadLetterSinkName)
```
Comment (matzew) [inline test/e2e_new/dls_extensions_test.go:157]: while we do use the "Context" from the `deadLetterSinkName` that received the event, we still want the actual "error destination" to be the URL from the original sink

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Pull request merged**